pkg/xds/server/callbacks/dataplane_callbacks_test.go

```diff
@@ -64,44 +59,90 @@ var _ = Describe("Dataplane Callbacks", func() {
 
-		// then only OnProxyReconnected should be called
+		// then OnProxyReconnected should not be called twice
 		Expect(countingCallbacks.OnProxyConnectedCounter).To(Equal(1))
-		Expect(countingCallbacks.OnProxyReconnectedCounter).To(Equal(1))
 
 		// when first stream is closed
 		callbacks.OnStreamClosed(1, node)
 
-		// then OnProxyDisconnected should not yet be called
-		Expect(countingCallbacks.OnProxyDisconnectedCounter).To(Equal(0))
+		// then OnProxyDisconnected should be called
+		Expect(countingCallbacks.OnProxyDisconnectedCounter).To(Equal(1))
 
 		// when last stream is closed
 		callbacks.OnStreamClosed(2, node)
 
-		// then OnProxyDisconnected should be called
+		// then OnProxyDisconnected should not be called twice
 		Expect(countingCallbacks.OnProxyDisconnectedCounter).To(Equal(1))
 	})
+
+	It("should reject new stream when the cleanup is not complete", func() {
+		// setup
+		cleanupCost := 2 * time.Second
+		cleanupStarted := make(chan struct{})
+		tracker := NewDataplaneSyncTracker(func(key core_model.ResourceKey) util_xds_v3.Watchdog {
+			return WatchdogFunc(func(ctx context.Context) {
+				<-ctx.Done()
+				cleanupStarted <- struct{}{}
+				<-time.After(cleanupCost)
+			})
+		})
+		callbacks := util_xds_v3.AdaptCallbacks(DataplaneCallbacksToXdsCallbacks(tracker))
+
+		// when one stream for backend-01 is connected and request is sent
+		streamID1 := int64(1)
+		streamID2 := int64(2)
+		streamID3 := int64(3)
+		err := callbacks.OnStreamOpen(context.Background(), streamID1, "")
+		Expect(err).ToNot(HaveOccurred())
+		node := &envoy_core.Node{Id: "default.backend-01"}
+		err = callbacks.OnStreamRequest(streamID1, &envoy_sd.DiscoveryRequest{Node: node})
+		Expect(err).ToNot(HaveOccurred())
+
+		// and when a new stream from backend-01 tries to connect quickly before the cleanup is complete
+		go callbacks.OnStreamClosed(streamID1, node)
+		<-cleanupStarted
+		<-time.After(cleanupCost - time.Second)
+
+		err = callbacks.OnStreamOpen(context.Background(), streamID2, "")
+		Expect(err).ToNot(HaveOccurred())
+		err = callbacks.OnStreamRequest(streamID2, &envoy_sd.DiscoveryRequest{Node: node})
+
+		// then it should be rejected
+		Expect(err).To(HaveOccurred())
+		Expect(err.Error()).To(ContainSubstring("already an active stream"))
+		go callbacks.OnStreamClosed(streamID2, node)
+
+		// and when a new stream from backend-01 tries to connect quickly after the cleanup is complete
+		<-time.After(cleanupCost)
+		// and when the third stream from backend-01 is connected after the first active stream closed and request is sent
+		err = callbacks.OnStreamOpen(context.Background(), streamID3, "")
+		Expect(err).ToNot(HaveOccurred())
+		err = callbacks.OnStreamRequest(streamID3, &envoy_sd.DiscoveryRequest{Node: node})
+		Expect(err).ToNot(HaveOccurred())
+
+		// when other stream is closed and the third stream is open
+		go callbacks.OnStreamClosed(streamID3, node)
+	})
```

Comment:
Do we need this test? I feel that this exact behavior is tested in pkg/xds/server/callbacks/dataplane_sync_tracker_test.go already

Reply:
Right, it's tested over there. I'm thinking if it's better to test it here and remove the test in that file. The case in that file named `should start only one watchdog per dataplane`, but the `dataplane_sync_tracker.go` itself does not limit it. To me, this feature is provided by this `dataplane_callbacks.go`

Reply:
Removed